[dtensor][2/N] more tensor ops to use strategy propagation

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101203

As titled, this PR adapts a few more tensor ops to use strategy based
sharding prop

wanchaol committed:
[dtensor][2/N] more tensor ops to use strategy propagation

As titled, this PR adapts a few more tensor ops to use strategy based
sharding prop

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][2/N] more tensor ops to use strategy propagation"


As titled, this PR adapts a few more tensor ops to use strategy based
sharding prop

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][2/N] more tensor ops to use strategy propagation"


As titled, this PR adapts a few more tensor ops to use strategy based
sharding prop

[ghstack-poisoned]

XilunWu commented on torch/distributed/_tensor/ops/tensor_ops.py:
```diff
@@ -203,85 +198,10 @@ def replicate_tensor_dim(
-@register_prop_rule(aten.slice_scatter.default)
-def prop_slice_scatter(op_schema: OpSchema) -> OutputSharding:
+@register_op_strategy(aten.slice_scatter.default)
+def gen_slice_scatter_strategy(
```
question: can we merge the common logic with `gen_slice_strategy`?

XilunWu commented on torch/distributed/_tensor/ops/tensor_ops.py:
```diff
@@ -290,60 +210,47 @@ def prop_slice_scatter(op_schema: OpSchema) -> OutputSharding:
+        for arg_strategy in input_strategy.strategies:
+            arg_spec = arg_strategy.output_spec
+            replicate_spec = DTensorSpec(
+                mesh, replicate_tensor_dim(arg_spec.placements, dim=slice_dim)
```
the difference between `replicate_tensor_dim` and `unshard_tensor_dim` is, `unshard_tensor_dim` only replicate when the dim is `Shard` while `replicate_tensor_dim` replicate when the dim is either `Shard` or `_Partial`. What is the reason doing this?

wanchaol replied:
`replicate_tensor_dim` means if tensor dim is sharded we replicate it, also not allowing partial, `unshard` tensor dim only checks if it's sharded but not check partial, this is to deal with cases are not accepting partial, i.e. `slice_scatter` not accepting partial

XilunWu commented on torch/distributed/_tensor/ops/utils.py:
```diff
@@ -69,3 +70,33 @@ def normalize_dims(dims: Union[int, Sequence[int]], ndim: int) -> Sequence[int]:
+            shards_map[shard_dim] *= spec.mesh.size(i)
+
+    for i, dim_size in enumerate(shape):
+        # TODO: maybe we should determine is_sharable base on the
+        #      whether it's evenly sharded or not
```
Is the indentation correct here? Can polish the `TODO` information:
"is_sharable" -> "is_shardable"
"base" -> "based"
"on the whether" -> "on whether"

XilunWu reviewed (approved):
lgtm! some nits and questions.

fduwjj reviewed (commented):
Do we have a design or RFC for the op strategy propagation?

fduwjj commented on torch/distributed/_tensor/ops/tensor_ops.py:
```diff
@@ -290,60 +210,47 @@ def prop_slice_scatter(op_schema: OpSchema) -> OutputSharding:
+
+@register_op_strategy(aten._local_scalar_dense.default)
+def replica_only_strategy(
+    node: Node, mesh: DeviceMesh, node_to_strategy: Dict[Node, StrategyType]
```
ditto.

fduwjj commented on torch/distributed/_tensor/ops/utils.py:
```diff
@@ -69,3 +70,33 @@ def normalize_dims(dims: Union[int, Sequence[int]], ndim: int) -> Sequence[int]:
+    for i, dim_size in enumerate(shape):
+        # TODO: maybe we should determine is_sharable base on the
+        #      whether it's evenly sharded or not
+        if dim_size < shards_map[i]:
+            return False
```
Is this to avoid cases when one rank has no local tensor?

fduwjj reviewed (commented):
Just another n00b question, can eager mode also leverage this work or we definitely need to first have a fx trace graph first?

fduwjj reviewed (commented):
Just another n00b question, can eager mode also leverage this work or we definitely need to first have a fx trace graph first?

wanchaol committed:
Update on "[dtensor][2/N] more tensor ops to use strategy propagation"


As titled, this PR adapts a few more tensor ops to use strategy based
sharding prop

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][2/N] more tensor ops to use strategy propagation"


As titled, this PR adapts a few more tensor ops to use strategy based
sharding prop

[ghstack-poisoned]

wanchaol commented:
> Just another n00b question, can eager mode also leverage this work or we definitely need to first have a fx trace graph first?

The fx graph is on the op level, used to propagate the tensor metadata and sharding, both eager and graph could leverge this

wanchaol committed:
Update on "[dtensor][2/N] more tensor ops to use strategy propagation"


As titled, this PR adapts a few more tensor ops to use strategy based
sharding prop

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][2/N] more tensor ops to use strategy propagation"


As titled, this PR adapts a few more tensor ops to use strategy based
sharding prop

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
